pkg/cloud/externalconn/record.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+// externalConnectionRecord is a reflective representation of a row in a
+// system.external_connections table.
+//
+// Each field in this struct has a tag specifying the column name in the
+// system.external_connections table containing the data for the field. Do not
+// manipulate these fields directly, use methods in the ExternalConnection.
+type externalConnectionRecord struct {
+	ConnectionName    string                         `col:"connection_name"`
+	ConnectionType    string                         `col:"connection_type"`
+	ConnectionDetails connectionpb.ConnectionDetails `col:"connection_details"`
+}
+
+// ExternalConnection is a representation of an External Connection.
+//
+// This struct can marshal/unmarshal changes made to the underlying
+// system.external_connections table.
+type ExternalConnection struct {
+	// The "record" for this external connection.  Do not access this field
+	// directly (except in tests); Use Get/Set methods on ExternalConnection
+	// instead.
+	rec externalConnectionRecord
+
+	// Set of changes to this external connection that need to be persisted.
+	dirty map[string]struct{}
+}
+
+// NewExternalConnection creates and initializes an ExternalConnection.
+func NewExternalConnection() *ExternalConnection {
+	return &ExternalConnection{
+		dirty: make(map[string]struct{}),
+	}
+}
+
+// SetConnectionName updates the connection name.
+func (e *ExternalConnection) SetConnectionName(name string) {
+	e.rec.ConnectionName = name
+	e.markDirty("connection_name")
+}
+
+// SetConnectionType updates the connection_type.
+func (e *ExternalConnection) SetConnectionType(connectionType string) {
+	e.rec.ConnectionType = connectionType
+	e.markDirty("connection_type")
+}
+
+// SetConnectionDetails updates the connection_details.
+func (e *ExternalConnection) SetConnectionDetails(details connectionpb.ConnectionDetails) {
+	e.rec.ConnectionDetails = details
+	e.markDirty("connection_details")
+}
+
+// datumToNative is a helper to convert tree.Datum into Go native types.  We
+// only care about types stored in the system.external_connections table.
+func datumToNative(datum tree.Datum) (interface{}, error) {
+	datum = tree.UnwrapDOidWrapper(datum)
+	if datum == tree.DNull {
+		return nil, nil
+	}
+	switch d := datum.(type) {
+	case *tree.DString:
+		return string(*d), nil
+	case *tree.DBytes:
+		return []byte(*d), nil
+	}
+	return nil, errors.Newf("cannot handle type %T", datum)
+}
+
+// InitFromDatums initializes the receiver based on datums and column names.
+func (e *ExternalConnection) InitFromDatums(
```

Comment:
I feel like I've seen this code before :) It's copy-pasta'ed from schedule jobs, right? any chance we could add or use a generic library?

Reply:
Heavily inspired is the term I use 😋  Yeah ~20ish lines are shareable, but since the method is responsible for init'ing the receiver from datums there is logic specific to the scheduled job or external connection object intertwined. For example, the scheduled jobs has a concept of an `owner` column that needs special casing. Furthermore the reflection we perform at the top is specific to the record stored in the receiver:

`record := reflect.ValueOf(&j.rec).Elem()`

This is one of those cases where slight duplication might make it easier to maintain/understand but if you feel strongly about it I can take another stab.

Reply:
Generics (not available) might be an answer... though I doubt it... Regardless, it seems that maybe you could just... hard code ser/de stuff -- this object only has 3 fields...

Reply:
Up to you :)

Reply:
I do not feel strongly! That makes sense, whichever we seems best to you is 👍